processor/src/advice/dsa.rs

```diff
@@ -0,0 +1,68 @@
+    utils::Deserializable,
+};
+
+/// Gets as input a vector containing an expanded public key and its associated secret key, and a
+/// word representing a message and outputs a vector of values to be pushed onto the advice stack.
+/// The values are the ones required for a Falcon signature verification inside the VM and they are:
+///
+/// 1. The nonce represented as 8 field elements.
+/// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+/// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+/// 4. The product of the above two polynomials in the ring of polynomials with coefficients
+/// in the Miden field.
+///
+/// # Errors
+/// Will return an error if either:
+/// - The advice map does not contain an entry with key PK.
+/// - The advice map entry under key PK is not a vector of the expected length.
+///
+/// The function only generates non-deterministic input that is required for the Falcon verification
+/// procedure inside the VM and as such does interact with the VM only through the advice provider.
+pub fn falcon_sign(pk_sk: &[Felt], msg: Word) -> Result<Vec<Felt>, ExecutionError> {
+    // Create the corresponding key pair
+    let mut key_pair_bytes = Vec::with_capacity(pk_sk.len());
+    for element in pk_sk {
+        let value = element.as_int();
+        if value > u8::MAX as u64 {
+            return Err(ExecutionError::MalformedSignatureKey("RPO Falcon512"));
+        }
+        key_pair_bytes.push(value as u8);
+    }
+
+    let key_pair = KeyPair::read_from_bytes(&key_pair_bytes)
+        .map_err(|_| ExecutionError::MalformedSignatureKey("RPO Falcon512"))?;
+
+    // We can now generate the signature
+    let sig = key_pair
+        .sign(msg)
+        .map_err(|_| ExecutionError::FailedSignatureGeneration("RPO Falcon512"))?;
+
+    // The signature is composed of a nonce and a polynomial s2
+
+    // The nonce is represented as 8 field elements.
+    let nonce = sig.nonce();
+
+    // We convert the signature to a polynomial
+    let s2: Polynomial = sig.sig_poly();
+
+    // We also need in the VM the expanded key corresponding to the public key the was provided
+    // via the operand stack
+    let h: Polynomial = sig.pub_key_poly();
+
+    // Lastly, for the probabilistic product routine that is part of the verification procedure,
+    // we need to compute the product of the expanded key and the signature polynomial in
+    // the ring of polynomials with coefficients in the Miden field.
+    let pi = Polynomial::mul_modulo_p(&h, &s2);
+
+    // We now push the nonce, the expanded key, the signature polynomial, and the product of the
+    // expanded key and the signature polynomial to the advice stack.
+    let mut result: Vec<Felt> = nonce.to_vec();
+    result.extend(h.inner().iter().map(|a| Felt::new(*a as u64)));
+    result.extend(s2.inner().iter().map(|a| Felt::new(*a as u64)));
+    result.extend(pi.iter().map(|a| Felt::new(*a)));
```

Comment:
very minor nit: instead of `Felt::new()` we could use `Felt::from()` here and avoid the need for upcasting to `u64`.

Reply:
Fixed, thank you!